tt_eager/tt_dnn/op_library/composite/composite_ops.cpp

```diff
@@ -834,6 +834,26 @@ Tensor addcdiv(
     return operation::decorate_as_composite(__func__, _addcdiv)(input_a, input_b, input_c, value, output_mem_config);
 }
 
+Tensor _binary_div(
+    const Tensor& input_a,
+    const Tensor& input_b,
+    const MemoryConfig& output_mem_config) {
+    Tensor result = mul(input_a, recip(input_b, output_mem_config), std::nullopt, output_mem_config);
+    Tensor t_inf = full_like(input_a, std::numeric_limits<float>::infinity(), output_mem_config);
+    Tensor t_nan = full_like(input_a, std::nanf(""), output_mem_config);
+    return where(eqz(input_b, output_mem_config),
+                where(eqz(input_a, output_mem_config),
+                    t_nan,
+                    mul(t_inf, sign(input_a, output_mem_config), std::nullopt, output_mem_config), output_mem_config),
+                result, output_mem_config);
```

Comment:
I would prefer we at least expose just the mul with recip as its own div op (our eltwise binary implementation supports fusing unary ops to prescale binary inputs). This support for handling nan adds a lot of extra ops and overhead, whereas if the user knows / doesn't care about nans, then having this exposed separately would be much better for perf since it would only be a single op. Not sure what your opinion on this is @eyonland ? This might be true for other ops as well that we could potentially expose a path like this.

We should also review this code for handling nans that is being added everywhere to make it more performant, as well as commonize the logic instead of copy pasting similar code everywhere

Reply:
Hi @tt-aho , @eyonland 
I tried implementing using Unary chain method
```
     case BinaryOpType::BINARY_DIV:
            defines.merge(eltwise_unary_op_utils::get_defines(UnaryOpType::RECIP, std::nullopt, "PRE_IN1_0"));
            op_name = "mul_tiles";
            op_code = "2";
            break;
```

On using this method, the test are failing (assertion error). This is due to hardware restrictions related to nan  as  discussed in issue #4409 . Below shared is the result comparison for Binary_div on implementing in composite structure and unary chain method

- Composite structure ( Current implementation ) - Tests PASSED
<img width="1119" alt="Screenshot 2024-04-11 at 3 44 25 PM" src="https://github.com/tenstorrent-metal/tt-metal/assets/138196495/d26c2101-9ded-46d6-9cc9-36f247f81ffb">

- Unary chain method - Tests FAILED
<img width="1164" alt="Screenshot 2024-04-11 at 3 47 01 PM" src="https://github.com/tenstorrent-metal/tt-metal/assets/138196495/9035409e-cdbb-4e83-b3ee-b40c47ec089b">

Would like your inputs on this.
Thanks.

Reply:
@VirdhatchaniKN my comment did not say to replace the composite method as I understand that the non-composite version would not produce correct output when denominator is 0. It meant that if user knows/can guarantee the denominator is not 0, then they can avoid a lot of overhead since the current logic to handle nans adds >= 6 additional ops plus overhead of creating the 2 full tensors.